critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -233,49 +236,83 @@ public void run() {
     /**
      * Go through contents of urlIterator, attempting to sync the artifact.
      * @param urlIterator
-     * @param artifact
-     * @param sa
      * @return success: return Storage Metadata from first successful sync
      *         fail: return null if all URLs failed | throw if failure is fatal (internal)
      * @throws StorageEngageException
      * @throws InterruptedException
      * @throws WriteException
+     * @throws IllegalArgumentException
      */
-    private StorageMetadata syncArtifact(Iterator<URL> urlIterator, URI artifact, StorageAdapter sa)
-        throws StorageEngageException, InterruptedException, WriteException {
+    private StorageMetadata syncArtifact(Iterator<URL> urlIterator)
+        throws StorageEngageException, InterruptedException, WriteException, IllegalArgumentException {
 
         StorageMetadata storageMeta = null;
 
-        while (urlIterator.hasNext()) {
-            URL u = urlIterator.next();
-            log.debug("trying " + u);
+        if (urlIterator.hasNext()) {
+            while (urlIterator.hasNext()) {
+                URL u = urlIterator.next();
+                log.debug("trying " + u);
 
-            try {
-                ByteArrayOutputStream dest = new ByteArrayOutputStream();
-                HttpGet get = new HttpGet(u, dest);
-                get.prepare();
-
-                NewArtifact a = new NewArtifact(artifact);
-                storageMeta = sa.put(a, get.getInputStream());
-                log.debug("storage meta returned: " + storageMeta.getStorageLocation());
-                return storageMeta;
-
-            } catch (WriteException wre) {
-                // IOException will capture this if not explicitly caught
-                log.info("write error for storage adapter put: " + wre.getMessage());
-                throw wre;
-            } catch (TransientException | IOException te) {
-                // ReadException will be caught under the IOException
-                // could be prepare or put throwing this error
-                // will move to next url
-                log.info("transient exception." + te.getMessage());
-
-            } catch (ResourceNotFoundException | ResourceAlreadyExistsException  badURL) {
-                log.info("removing URL " + u + " from list: " + badURL.getMessage());
-                urlIterator.remove();
+                try {
+                    ByteArrayOutputStream dest = new ByteArrayOutputStream();
+                    HttpGet get = new HttpGet(u, dest);
+                    get.prepare();
 
+                    // Check content checksum and length to verify this
+                    // artifact can be synced currently
+                    Artifact curArtifact = this.artifactDAO.get(this.artifactID);
+
+                    if (curArtifact.storageLocation != null) {
+                        // artifact has been acted on since this FileSyncJob
+                        // instance started and it was null
```

Comment:
I would have agonised about what exception to throw here. I'm generally reluctant to use the exception mechanism for flow control but we already used null to mean "nothing accomplished but not fatal either" and this scenario does need to terminate the job.

`IllegalArgumentException` typically means the input (iterator) is invalid input for this method and could be thrown by other code in here (because it gets thrown a lot by constructors). 

Since the input was fine and the scenario is that the state of the system as a whole is not as expected, I would have ended up chosing `IllegalStateException` and had a separate catch for it in run() -- because when we go back through here and make it log in json format for ingestion  into ELK, we'll have to distinguish whether this "job" was success = true|false and I would want that logged as success=true when this happens. 

It's admittedly seems pretty pedantic :-)

Reply:
I did agonise to a point - but also knew you'd tell me if I got it wrong. :)